src/connector/src/source/external.rs

```diff
@@ -0,0 +1,590 @@
+        format!("`{}`", table_name.table_name)
+    }
+
+    fn current_binlog_offset(&self) -> Self::BinlogOffsetFuture<'_> {
+        async move {
+            let mut conn = self
+                .pool
+                .get_conn()
+                .await
+                .map_err(|e| ConnectorError::Connection(anyhow!(e)))?;
+
+            let sql = "SHOW MASTER STATUS".to_string();
+            let mut rs = conn.query::<mysql_async::Row, _>(sql).await?;
+            let row = rs
+                .iter_mut()
+                .exactly_one()
+                .map_err(|e| ConnectorError::Internal(anyhow!("read binlog error: {}", e)))?;
+
+            Ok(BinlogOffset::MySql(MySqlOffset {
+                filename: row.take("File").unwrap(),
+                position: row.take("Position").unwrap(),
+            }))
+        }
+    }
+
+    fn parse_binlog_offset(&self, offset: &str) -> ConnectorResult<BinlogOffset> {
+        Ok(BinlogOffset::MySql(MySqlOffset::parse_str(offset)?))
+    }
+
+    fn snapshot_read(
+        &self,
+        table_name: SchemaTableName,
+        start_pk: Option<OwnedRow>,
+        primary_keys: Vec<String>,
+    ) -> BoxStream<'_, ConnectorResult<OwnedRow>> {
+        self.snapshot_read_inner(table_name, start_pk, primary_keys)
+    }
+}
+
+impl MySqlExternalTableReader {
+    pub fn new(properties: HashMap<String, String>, rw_schema: Schema) -> ConnectorResult<Self> {
+        let config = serde_json::from_value::<ExternalTableConfig>(
+            serde_json::to_value(properties).unwrap(),
+        )
+        .map_err(|e| {
+            ConnectorError::Config(anyhow!("fail to extract mysql connector properties: {}", e))
+        })?;
+
+        let database_url = format!(
+            "mysql://{}:{}@{}:{}/{}",
+            config.username, config.password, config.host, config.port, config.database
+        );
+        let pool = mysql_async::Pool::from_url(database_url)?;
+        Ok(Self {
+            pool,
+            config,
+            rw_schema,
+        })
+    }
+
+    pub async fn disconnect(&self) -> ConnectorResult<()> {
+        self.pool.clone().disconnect().await?;
+        Ok(())
+    }
+
+    #[try_stream(boxed, ok = OwnedRow, error = ConnectorError)]
+    async fn snapshot_read_inner(
+        &self,
+        table_name: SchemaTableName,
+        start_pk: Option<OwnedRow>,
+        primary_keys: Vec<String>,
+    ) {
+        let order_key = primary_keys.iter().join(",");
+        let sql = if start_pk.is_none() {
+            format!(
+                "SELECT * FROM {} ORDER BY {}",
+                self.get_normalized_table_name(&table_name),
+                order_key
+            )
+        } else {
+            let filter_expr = Self::filter_expression(&primary_keys);
+            format!(
+                "SELECT * FROM {} WHERE {} ORDER BY {}",
+                self.get_normalized_table_name(&table_name),
+                filter_expr,
+                order_key
+            )
+        };
```

Comment:
I am considering whether we should add a `limit` to these SQLs because some drivers like jdbc will consume all the remaining rows when we try to close the connection. Could you please check what is the behavior of `mysql_async`?